### ISSUE TITLE
Hokuto no Ken Raoh Gaiden Ten no Haoh black screen

### ISSUE THREAD

**Issue by daniel229** [hrydgard/ppsspp#7932]
Accidentally fixed for D3D9 in https://github.com/hrydgard/ppsspp/pull/7920 ,Broken by the last commit https://github.com/hrydgard/ppsspp/commit/488c62e5349d3d4bb8bc82aecb173deb71fe2843

GEdebugger
![01](https://cloud.githubusercontent.com/assets/3481559/9567444/24363de2-4f5e-11e5-961a-0611b37ebf3b.png)

more tabs
https://gist.github.com/daniel229/90091277dabb78e0b1bc

**Comment by hrydgard** [comment]
Also broke Wipeout menus, probably related. I will look into it...

**Comment by daniel229** [comment]
With https://github.com/hrydgard/ppsspp/commit/08b340f423c5c4a09653cca345f4fd43297cffaa ,still black screen,but the loading screen displayed.

**Comment by unknownbrackets** [comment]
Does anything change if you modify this in GLES_GPU.cpp:

```c++
	if ((!gl_extensions.IsGLES || gl_extensions.GLES3) && PSP_CoreParameter().compat.flags().PixelDepthRounding) {
		features |= GPU_ROUND_FRAGMENT_DEPTH_TO_16BIT;
```

To just:

```c++
	if ((!gl_extensions.IsGLES || gl_extensions.GLES3)) {
		features |= GPU_ROUND_FRAGMENT_DEPTH_TO_16BIT;
```

Or add it to the ini to see, either way.  Just wondering if depth values change things.

-[Unknown]

**Comment by daniel229** [comment]
Nothing change

**Comment by unknownbrackets** [comment]
It's interesting that the Z is `-1.562500` in the vertices list.  Clip is enabled too, but depth write is disabled.

If you go back to a working build, does the Z column in the vertices list show a different value?

-[Unknown]

**Comment by daniel229** [comment]
This is the working build, Z is the same.

![1](https://cloud.githubusercontent.com/assets/3481559/12008444/b7da446a-ac76-11e5-90fd-e846123fbc71.png)

https://gist.github.com/daniel229/37da3fa223bfe87bb7b8

**Comment by unknownbrackets** [comment]
If you revert just the StateMappingDX9 part of 488c62e5349d3d4bb8bc82aecb173deb71fe2843, does it work again?

-[Unknown]

**Comment by daniel229** [comment]
No,it need to revert https://github.com/hrydgard/ppsspp/commit/7b0a31ade539f9e1e90801cf79a2b12deb5d2e00

**Comment by unknownbrackets** [comment]
Oh okay, but if you remove that offset entirely, it draws fine?

In OpenGL, when it's about to draw the things that are missing in Direct3D9, can you flip OFF the "depth write disable"?  I want it to write the depth value.  After it's drawn, you can switch to the "Depth" tab, and hover over the areas it drew - I'm just wondering what the value comes out to.

-[Unknown]

**Comment by daniel229** [comment]
No,it it need to revert all things before https://github.com/hrydgard/ppsspp/commit/7b0a31ade539f9e1e90801cf79a2b12deb5d2e00

like this?
![01](https://cloud.githubusercontent.com/assets/3481559/12008855/fe61a7a6-ac95-11e5-9445-ce7cdfb68419.png)

**Comment by unknownbrackets** [comment]
Yes, I think, but make sure "Depth write disable" shows 0 before the draw (double click it.)  And after the draw, move the mouse over the part that WAS red, it should replace the "Depth: 0x00088000 (1920x1080)" text with the depth value under your mouse cursor.

-[Unknown]

**Comment by daniel229** [comment]
But Depth write disable is always 0.
![04](https://cloud.githubusercontent.com/assets/3481559/12009271/0800f102-acac-11e5-9d95-9b7f349cfd95.png)

**Comment by dan83sama** [comment]
In software mode menus show prefect, and Z = 1,

http://i.imgur.com/j6TgCG5.png

**Comment by unknownbrackets** [comment]
Hmm.  Has this improved at all in the latest git build?  We've made some depth changes.

-[Unknown]

**Comment by daniel229** [comment]
No,still the same.

**Comment by unknownbrackets** [comment]
Does this change make it render?

https://github.com/hrydgard/ppsspp/issues/8424#issuecomment-172410462

-[Unknown]

**Comment by daniel229** [comment]
No.

**Comment by unknownbrackets** [comment]
This has clipping enabled, which I've recently discovered does some weird things with depth.  Does #8461 help?

-[Unknown]

**Comment by daniel229** [comment]
No,it doesn't.

**Comment by LunaMoo** [comment]
I tried to find lowest value it works for the hack I described [here](http://forums.ppsspp.org/showthread.php?tid=19091&pid=118216#pid118216) and z /= 1.000031 was enough. Don't think there's anything special with this value, through it seems weird that cutting depth by just soo little makes a difference.

**Comment by unknownbrackets** [comment]
Actually, I think I know why this is happening but it's annoying / tricky to fix.  But I'm not sure why it shows -1.5 - it could be I'm wrong.

When rendering, there are two options for depth:
1. Clamp depth; if it's outside the range of [0.0, 65535.0] after final transforms etc. then clamp it to that range.  Never clip based on Z.

2. Clip depth, not rendering parts of polygons with depth outside [-0.5, 65536.0].  If the depth is between [-0.5, 0.0) then clamp to 0.  If the depth is between (65535.0, 65536.0] then clamp to 65535.0.

We do NEITHER of these accurately.  Instead, we clip anything outside of [0.0, 65535.0].  Clamping isn't something modern devices like - in general they don't want Z to change after the vertex shader, because if it doesn't change, they can perform early Z tests.

I could be wrong, but my guess is that the PSP (when clipping) is performing early Z tests, but they're inaccurate, and then it falls back to clamping, which it probably always does.  It then also has a per-fragment Z test, which allows discarding outside a configurable range as well (which seems mostly accurate.)

Anyway, this game has clipping enabled.  And it stopped working when we stopped incorrectly applying a -0.5/65536 offset to the Z.  I think Z is falling just outside [0, 65535.0], although again I'm not sure why it shows as -1.5 in the debugger.

-[Unknown]

**Comment by benderscruffy** [comment]
v1.5.4-933-g386c9d4fc
still present

**Comment by Saramagrean** [comment]
comfirm is fixed in v1.6.3 & v1.7.1 latest git build,

**Comment by Saramagrean** [comment]
Confirm is fixed in v1.6.3 & v1.7.1 latest git build. except D3D9 is blackscreen.

**Comment by unknownbrackets** [comment]
Does this work properly in Vulkan or Direct3D 11?  It probably won't work in OpenGL or Direct3D 9 as we don't clamp depth there.  It may not work on all Vulkan devices.

-[Unknown]

**Comment by benderscruffy01** [comment]
goes ingame no problem using  Direct3D 11 backend
![hokuto](https://user-images.githubusercontent.com/48821663/54874242-251ca680-4e22-11e9-8afb-f4f82f4409e2.PNG)

**Comment by nassau-tk** [comment]
This issue happen in android(OpenGL) yet.
But,Vulkan is drawing collectly.

I got the dump file.(OpenGL) 
[ULJM05404_0001.zip](https://github.com/hrydgard/ppsspp/files/4318508/ULJM05404_0001.zip)

v1.9.3-465

**Comment by Gamemulatorer** [comment]
> I got the dump file.(OpenGL)
> [ULJM05404_0001.zip](https://github.com/hrydgard/ppsspp/files/4318508/ULJM05404_0001.zip)
> 
> v1.9.3-561

Using the dump above.
### HW rendering
![Screenshot_20200604-170410](https://user-images.githubusercontent.com/37603562/83737423-8f1b1780-a685-11ea-937f-5c7e9bac79d4.png)

### SW rendering
![Screenshot_20200604-170356](https://user-images.githubusercontent.com/37603562/83737442-96422580-a685-11ea-85fd-4348e29b0b70.png)

**Comment by Panderner** [comment]
I used save state for in-game, in OpenGL 3D Models does display but not the 2d ones.
![Screenshot_2020-12-22-19-06-24-183_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/102882727-cc13f380-4489-11eb-8d70-349a02146c19.jpg)

Here's a GE dump:
[GE dump.zip](https://github.com/hrydgard/ppsspp/files/5729497/GE.dump.zip)

**Comment by Gamemulatorer** [comment]
This issue can also reproduce on Vulkan.
![Screenshot_2021-09-05-19-46-49-651_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/132125595-caac8c04-add7-4ca2-83b2-161c126113fe.jpg)
[recording.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7111513/recording.ppdmp.zip)

**Comment by Gamemulatorer** [comment]
This game hang on SW Rendering during the fight.
![Screenshot_2021-09-05-22-11-11-689_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/132130121-34f15cf5-7f0a-42bd-897e-33a33c6f5b0a.jpg)

[Hokuto_No_Ken_Raoh_Gaiden.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7111750/Hokuto_No_Ken_Raoh_Gaiden.ppdmp.zip)

### Debug
```
<<< log_count = 1503 >>>
[09-03 23:09:55.466 11356:11356 F/DEBUG]
pid: 11163, tid: 11286, name: AndroidRender  >>> org.ppsspp.ppsspp <<<

[09-03 23:09:55.727 11356:11356 F/DEBUG]
      #04 pc 00000000006b84f0  /data/app/~~SlE13pXGaCtZaOaCh6oLFQ==/org.ppsspp.ppsspp-2lwBVG9AWLNVy6z3AEzsTA==/lib/arm64/libppsspp_jni.so (HandleAssert(char const*, char const*, int, char const*, char const*, ...)+248) (BuildId: a44dc90521dc05b22af687988046c386f6b18466)

[09-03 23:09:55.727 11356:11356 F/DEBUG]
      #05 pc 000000000051c310  /data/app/~~SlE13pXGaCtZaOaCh6oLFQ==/org.ppsspp.ppsspp-2lwBVG9AWLNVy6z3AEzsTA==/lib/arm64/libppsspp_jni.so (UtilityDialogInitialize(UtilityDialogType, int, int)+392) (BuildId: a44dc90521dc05b22af687988046c386f6b18466)

[09-03 23:09:55.727 11356:11356 F/DEBUG]
      #06 pc 000000000041c660  /data/app/~~SlE13pXGaCtZaOaCh6oLFQ==/org.ppsspp.ppsspp-2lwBVG9AWLNVy6z3AEzsTA==/lib/arm64/libppsspp_jni.so (PSPSaveDialog::Init(int)+936) (BuildId: a44dc90521dc05b22af687988046c386f6b18466)

[09-03 23:09:55.727 11356:11356 F/DEBUG]
      #07 pc 000000000051cecc  /data/app/~~SlE13pXGaCtZaOaCh6oLFQ==/org.ppsspp.ppsspp-2lwBVG9AWLNVy6z3AEzsTA==/lib/arm64/libppsspp_jni.so (BuildId: a44dc90521dc05b22af687988046c386f6b18466)

[09-03 23:09:55.727 11356:11356 F/DEBUG]
      #08 pc 000000000044c3ac  /data/app/~~SlE13pXGaCtZaOaCh6oLFQ==/org.ppsspp.ppsspp-2lwBVG9AWLNVy6z3AEzsTA==/lib/arm64/libppsspp_jni.so (CallSyscallWithoutFlags(HLEFunction const*)+28) (BuildId: a44dc90521dc05b22af687988046c386f6b18466)

[09-05 16:46:30.844 1371:2284 I/ActivityTaskManager]
START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.ppsspp.ppsspp/.PpssppActivity bnds=[44,1511][282,1771]} from uid 10491

[09-05 16:46:30.852 1371:2284 V/ActivityTaskManager]
Starting new activity ActivityRecord{5c530a2 u0 org.ppsspp.ppsspp/.PpssppActivity t310} in new task Task{9134133 #310 visible=true type=standard mode=fullscreen translucent=true A=10552:org.ppsspp.ppsspp U=0 StackId=310 sz=1}

[09-05 16:46:30.867 1371:1396 D/DexOptExtImpl]
org.ppsspp.ppsspp reason is unknown

[09-05 16:46:30.870 1371:1400 D/Boost]
hostingType=pre-top-activity, hostingName={org.ppsspp.ppsspp/org.ppsspp.ppsspp.PpssppActivity}, callerPackage=com.teslacoilsw.launcher, isSystem=false, isBoostNeeded=false.

[09-05 16:46:30.871 1371:1400 I/ActivityManager]
Start proc 8874:org.ppsspp.ppsspp/u0a552 for pre-top-activity {org.ppsspp.ppsspp/org.ppsspp.ppsspp.PpssppActivity} caller=com.teslacoilsw.launcher

[09-05 16:46:30.890 1371:1391 V/WindowManager]
Orientation start waiting for draw, mDrawState=DRAW_PENDING in Window{22edd23 mode=1 rootTaskId=310 u0 Splash Screen org.ppsspp.ppsspp}, surfaceController Surface(name=Splash Screen org.ppsspp.ppsspp)/@0x408287f

[09-05 16:46:31.402 1371:1400 D/Boost]
hostingType=content provider, hostingName={com.android.externalstorage/com.android.externalstorage.ExternalStorageProvider}, callerPackage=org.ppsspp.ppsspp, isSystem=false, isBoostNeeded=false.

[09-05 16:46:31.402 1371:1400 I/ActivityManager]
Start proc 8915:com.android.externalstorage/u0a79 for content provider {com.android.externalstorage/com.android.externalstorage.ExternalStorageProvider} caller=org.ppsspp.ppsspp
```

**Comment by unknownbrackets** [comment]
Yeah, I found a hang bug in the software renderer recently, will push a fix - hopefully it's the cause.

-[Unknown]

**Comment by Panderner** [comment]
Works fine for me in vulkan:
![Screenshot_2021-09-05-23-42-55-990_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/132133286-b7a7dd5d-57d7-4f06-8432-372d89b7c495.jpg)

**Comment by unknownbrackets** [comment]
See #12058 - this requires depth clamp.

This is actually a really good example of the behavior I describe in https://github.com/hrydgard/ppsspp/issues/12058#issuecomment-913225641, since all the points of the graphics it draws here have -1.5f depth, but that's post-viewport.

-[Unknown]

**Comment by Gamemulatorer** [comment]
> Works fine for me in vulkan:
> ![Screenshot_2021-09-05-23-42-55-990_org ppsspp ppsspp](https://user-images.githubusercontent.com/37503397/132133286-b7a7dd5d-57d7-4f06-8432-372d89b7c495.jpg)

IDK but it just blackscreen on my phone using vulkan I file a report https://github.com/hrydgard/ppsspp/issues/14796

**Comment by Panderner** [comment]
The title screen displays properly on PC using OpenGL:
![Ppsspp Screenshot 2021 09 13 - 00 37 56 07](https://user-images.githubusercontent.com/37503397/132995734-40d7b727-fc6c-45d7-b341-c6af94331d1a.png)

**Comment by Panderner** [comment]
At least it was fixed by #14042 only for desktop systems.

**Comment by Gamemulatorer** [comment]
I still got blackscreen on this game both opengl and vulkan. Im using the latest git build ppsspp.apk
:(
![Screenshot_2021-10-20-08-43-02-423_org ppsspp ppsspp](https://user-images.githubusercontent.com/37603562/138009428-696e416e-5a87-4ab7-a57b-805396e50dda.jpg)
[recording.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/7377366/recording.ppdmp.zip)

**Comment by nassau-tk** [comment]
Tested on v1.12.3.
My GalaxyS20 is working good on valkan.
OpenGL is coming the black screen.
Both, Meaning at title screen.

Vulkan
![Screenshot_20211020-145439_PPSSPP](https://user-images.githubusercontent.com/48179091/138036105-6268f2fb-2dfa-4cad-a0b4-512d4d603adb.jpg)

![Screenshot_20211020-144053_PPSSPP](https://user-images.githubusercontent.com/48179091/138035374-7bc554b4-7068-48cf-9378-9230eee29058.jpg)

**Comment by Panderner** [comment]
Umm yes this game requires depth clamp to display 2d graphics properly. Unfortunately mobile OpenGL and Mali Vulkan does not support depth clamp.

**Comment by sum2012** [comment]
v1.18.1-989 same issue in my phone

**Comment by Unknowncat1991** [comment]
D3D9 is out only issue here is Mali GPU's opengl lack of support Depth Clamp.

**Comment by Unknowncat1991** [comment]
> The game works correctly on my Samsung Galaxy A35:
> 
> ![Image](https://github.com/user-attachments/assets/f0a96d86-7851-4343-8b07-1aa5ebcd8c38)
> 
> Some Mali GPUs supports depth clamp which is supported on my phone.
> 
> ![Image](https://github.com/user-attachments/assets/4a84c92c-14d2-46ff-9950-ef7d63a0d67f)

If you switch to opengl still blackscreen

**Comment by Panderner** [comment]
What's the OpenGL feature that supports Depth Clamp? Mali GPUs lacks it.

**Comment by hrydgard** [comment]
Now that #21715 is in, this should work on all devices. Please re-open if not.
